posthog/warehouse/api/saved_query.py

```diff
@@ -330,6 +331,62 @@ def descendants(self, request: request.Request, *args, **kwargs) -> response.Res
 
         return response.Response({"descendants": descendants})
 
+    @action(methods=["POST"], detail=True)
+    def cancel(self, request: request.Request, *args, **kwargs) -> response.Response:
+        """Cancel a running saved query workflow."""
+        saved_query = self.get_object()
+
+        if saved_query.status != DataWarehouseSavedQuery.Status.RUNNING:
+            return response.Response(
+                {"error": "Cannot cancel a query that is not running"}, status=status.HTTP_400_BAD_REQUEST
+            )
+
+        temporal = sync_connect()
+        workflow_id = f"data-modeling-run-{saved_query.id.hex}"
+
+        try:
+            # Ad-hoc handling
+            try:
+                workflow_handle = temporal.get_workflow_handle(workflow_id)
+                if workflow_handle:
+                    async_to_sync(workflow_handle.cancel)()
+            except Exception:
+                logger.info("No ad-hoc workflow to cancel", workflow_id=workflow_id)
+
+            # Schedule handling
+            try:
+                scheduled_workflow_handle = temporal.get_schedule_handle(str(saved_query.id))
+                if scheduled_workflow_handle:
+                    desc = async_to_sync(scheduled_workflow_handle.describe)()
+                    recent_actions = desc.info.running_actions
+                    if len(recent_actions) > 0:
+                        most_recent_action = recent_actions[-1]
+                        if isinstance(most_recent_action, ScheduleActionExecutionStartWorkflow):
+                            workflow_id_to_cancel = most_recent_action.workflow_id
```

Comment:
This is currently the only concrete action class, but the type checker doesn't know this. So I think you may get complaints that `workflow_handle_to_cancel` is undefined.

Reply:
I had some issues with the typechecker here, but I think this seemed to resolve it for me as far as I can tell.